armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+
+    def close(self):
+        pass
+
+
+class NullWriter(Writer):
+    def write(self, record):
+        pass
+
+
+class PrintWriter(Writer):
+    def _write(self, name, batch, result):
+        print(f"Meter Record: name={name}, batch={batch}, result={result}")
+
+
+class LogWriter(Writer):
+    def __init__(self, log_level: str = "INFO"):
+        """
+        log_level - one of the uppercase log levels allowed by armory.logs.log
+        """
+        log.log(log_level, f"LogWriter set to armory.logs.log at level {log_level}")
+        self.log_level = log_level
+
+    def _write(self, name, batch, result):
+        log.log(
+            self.log_level, f"Meter Record: name={name}, batch={batch}, result={result}"
+        )
+
+
+class FileWriter(Writer):
+    def __init__(self, filepath):
+        self.filepath = filepath
+        self.file = open(self.filepath, "w")
+
+    def _write(self, name, batch, result):
+        record_json = json.dumps([name, batch, result])
+        # TODO: fix numpy output conversion issues
+        self.file.write(record_json + "\n")
+
+
+class ResultsWriter(Writer):
+    KEEP_MODES = ("first", "all", "last")
+
+    def __init__(self, sink, filepath, file_format="json", keep="all"):
+        # TODO: fix sink to callable function
+        self.sink = sink
+        self.filepath = filepath
+        self.file_format = file_format
+        if keep not in self.KEEP_MODES:
+            raise ValueError(f"keep {keep} not one of {self.KEEP_MODES}")
+        if keep != "all":
+            raise NotImplementedError(f"keep={keep}, use 'all'")
+        self.keep = keep
+        self.records = []
+        # TODO: checking
+
+    def set_filepath(self, filepath):
+        self.filepath = filepath
+
+    def _write(self, name, batch, result):
+        self.records.append((name, batch, result))
+
+    def collate_results(self):
+        """
+        Return a map from name to output, in original order.
+        """
+        output = {}
+        for name, batch, result in self.records:
+            if name not in output:
+                output[name] = []
+            output[name].append(result)
+        return output
+
+    def close(self):
+        output = self.collate_results()
+        with open(self.filepath, "w") as f:
+            json.dump(output, f)
+        # TODO: push to results dictionary?
+
+
+# GLOBAL CONTEXT METHODS #
+
+
+def get_context():
+    """
+    Get the context state object for the experimental procedure
+    """
+    global _CONTEXT
+    if _CONTEXT is None:
+        _CONTEXT = Context()
+    return _CONTEXT
+
+
+def get_probe(name: str = ""):
+    """
+    Get a probe with specified name, creating it if needed
+    """
+    if name != "" and not str.isidentifier(name):
+        raise ValueError(f"name {name} should be an identifier or the empty string")
+
+    if name not in _PROBES:
+        probe = Probe(name)
+        probe.set_sink(get_context())
```

Comment:
do sink and context need a different name if sink is just set to the context?

Reply:
I need to think more about this - in theory, you could send probes to anything that used the correct interface, not just the context. You could send it directly to an image exporter, to use your previous example.